File: Java/Data Structures & Algorithms Module - Java/Week5/Queue.java
/*
Queue.java
Billy Cussen
29/09/2020
*/

public class Queue<T> implements QueueInterface<T>{

    Node<T> first;
    Node<T> last;

    public Queue(){
        this.first = null;
        this.last = null;
    }

    @Override
    public void enQueue(T element){
        Node<T> oldLast = last;
        last = new Node<T>(element, null);
        if(isEmpty()){
            first = last;
        } else{
            oldLast.next = last;
        }
    }

    @Override
    public T deQueue() throws QueueEmptyException{
        if(isEmpty()){
            throw new QueueEmptyException();
        } else{
            T element = first.element;
            first = first.next;
            if(isEmpty()){
                last = null;
            }
            return element;
        }
    }

    @Override
    public T peek() throws QueueEmptyException{
        if(isEmpty()){
            throw new QueueEmptyException();
        } else{
            return first.element;
        }
    }

    @Override
    public boolean isEmpty(){
        return this.first == null;
    }

    @Override
    public int size(){
        return sizeHelper(first);
    }

    private int sizeHelper(Node<T> first){
        if(first == null){
            return 0;
        }
        return 1 + sizeHelper(first.next);
    }

    @Override
    public String toString(){
        int position = 0;
        String result = "***LINKED QUEUE TRAVERSAL***\n";
        Node<T> current = first;
        return toStringHelper(position, result, current);
    }

    private String toStringHelper(int position, String result, Node<T> current){
        if(current == null){
            return result;
        }
        result+="Position "+position+", Element: "+current.element+"\n";
        return toStringHelper(++position, result, current.next);
    }
}

File: Python/Data Structures & Algorithms Module - Python/Week5/Queue.py
"""
Queue.py
Billy Cussen
13/11/2020
"""
from Node import Node

class Queue:
    def __init__(self):
        self.first = None
        self.last = None

    def enQueue(self, elem):
        oldLast = self.last
        self.last = Node(elem, None)
        if self.isEmpty():
            self.first = self.last
        else:
            oldLast.next = self.last

    def deQueue(self):
        if self.isEmpty():
            raise Exception("The Queue is empty, can't deQueue!")
        else:
            elem = self.first.element
            self.first = self.first.next
            return elem

    def peek(self):
        if self.isEmpty():
            raise Exception("The Queue is empty, can't Peek!")
        else:
            return self.first.element

    def isEmpty(self):
        return self.first==None

    def getSize(self):
        return self.getSizeHelper(self.first)

    def getSizeHelper(self, current):
        if current == None:
            return 0
        return 1 + self.getSizeHelper(current.next)

    def toString(self):
        if self.isEmpty():
            raise Exception("The Queue is Empty, can't Traverse!")
        result, position = "***QUEUE TRAVERSAL***\n", 0
        return self.toStringHelper(self.first,result,position)

    def toStringHelper(self, current, result, position):
        if current == None:
            result += "***END***\n"
            return result
        result+="Position "+str(position)+", Element: "+str(current.element)+"\n"
        position+=1
        return self.toStringHelper(current.next,result,position)